// n8n-editor-ui/dist/assets/InsightsDashboard-BawwZ0Oo.js
const __vite__mapDeps=(i,m=__vite__mapDeps,d=(m.f||(m.f=["assets/InsightsPaywall-g3JcqAYL.js","assets/index-DD7Wuhk5.js","assets/index-DJ205bL0.css","assets/InsightsPaywall-OeXh6LeZ.css","assets/InsightsChartTotal-Cn386HO4.js","assets/index-BrjPYODL.js","assets/chartjs.utils-DTEY3u-V.js","assets/InsightsSummary-vMmKUcie.js","assets/InsightsSummary-W08OzElS.css","assets/InsightsChartFailed-DRi6JC3m.js","assets/InsightsChartFailureRate-BdFNsMpg.js","assets/InsightsChartTimeSaved-C12GZzSQ.js","assets/InsightsChartAverageRuntime-DWrrriYN.js","assets/InsightsTableWorkflows-DQQ1vvMz.js","assets/InsightsTableWorkflows-Us3j4UTO.css"])))=>i.map(i=>d[i]);
import { d as defineComponent, W as useRoute, gq as useInsightsStore, q as computed, bZ as defineAsyncComponent, I as watch, h as resolveComponent, i as createElementBlock, g as openBlock, j as createVNode, k as createBaseVNode, w as withCtx, l as createTextVNode, t as toDisplayString, m as unref, c as useI18n, e as createBlock, f as createCommentVNode, n as normalizeClass, F as Fragment, b2 as resolveDynamicComponent, aq as __vitePreload, _ as _export_sfc } from "./index-DD7Wuhk5.js";
import { I as InsightsSummary } from "./InsightsSummary-vMmKUcie.js";
const _sfc_main = /* @__PURE__ */ defineComponent({
  __name: "InsightsDashboard",
  props: {
    insightType: {}
  },
  setup(__props) {
    const InsightsPaywall = defineAsyncComponent(
      async () => await __vitePreload(() => import("./InsightsPaywall-g3JcqAYL.js"), true ? __vite__mapDeps([0,1,2,3]) : void 0)
    );
    const InsightsChartTotal = defineAsyncComponent(
      async () => await __vitePreload(() => import("./InsightsChartTotal-Cn386HO4.js"), true ? __vite__mapDeps([4,5,1,2,6,7,8]) : void 0)
    );
    const InsightsChartFailed = defineAsyncComponent(
      async () => await __vitePreload(() => import("./InsightsChartFailed-DRi6JC3m.js"), true ? __vite__mapDeps([9,5,1,2,6,7,8]) : void 0)
    );
    const InsightsChartFailureRate = defineAsyncComponent(
      async () => await __vitePreload(() => import("./InsightsChartFailureRate-BdFNsMpg.js"), true ? __vite__mapDeps([10,1,2,6,7,8,5]) : void 0)
    );
    const InsightsChartTimeSaved = defineAsyncComponent(
      async () => await __vitePreload(() => import("./InsightsChartTimeSaved-C12GZzSQ.js"), true ? __vite__mapDeps([11,1,2,6,7,8,5]) : void 0)
    );
    const InsightsChartAverageRuntime = defineAsyncComponent(
      async () => await __vitePreload(() => import("./InsightsChartAverageRuntime-DWrrriYN.js"), true ? __vite__mapDeps([12,1,2,6,7,8,5]) : void 0)
    );
    const InsightsTableWorkflows = defineAsyncComponent(
      async () => await __vitePreload(() => import("./InsightsTableWorkflows-DQQ1vvMz.js"), true ? __vite__mapDeps([13,1,2,7,8,14]) : void 0)
    );
    const props = __props;
    const route = useRoute();
    const i18n = useI18n();
    const insightsStore = useInsightsStore();
    const chartComponents = computed(() => ({
      total: InsightsChartTotal,
      failed: InsightsChartFailed,
      failureRate: InsightsChartFailureRate,
      timeSaved: InsightsChartTimeSaved,
      averageRunTime: InsightsChartAverageRuntime
    }));
    const getDefaultFilter = (query) => {
      const { time_span } = query;
      return {
        time_span: time_span ?? "7"
      };
    };
    const filters = computed(() => getDefaultFilter(route.query));
    const transformFilter = ({ id, desc }) => {
      const key = id;
      const order = desc ? "desc" : "asc";
      return `${key}:${order}`;
    };
    const fetchPaginatedTableData = ({
      page,
      itemsPerPage,
      sortBy
    }) => {
      const skip = page * itemsPerPage;
      const take = itemsPerPage;
      const sortKey = sortBy.length ? transformFilter(sortBy[0]) : void 0;
      void insightsStore.table.execute(0, {
        skip,
        take,
        sortBy: sortKey
      });
    };
    watch(
      () => filters.value.time_span,
      () => {
        if (insightsStore.isSummaryEnabled) {
          void insightsStore.summary.execute();
        }
        void insightsStore.charts.execute();
        void insightsStore.table.execute();
      },
      {
        immediate: true
      }
    );
    return (_ctx, _cache) => {
      const _component_N8nHeading = resolveComponent("N8nHeading");
      return openBlock(), createElementBlock("div", {
        class: normalizeClass(_ctx.$style.insightsView)
      }, [
        createVNode(_component_N8nHeading, {
          bold: "",
          tag: "h2",
          size: "xlarge"
        }, {
          default: withCtx(() => [
            createTextVNode(toDisplayString(unref(i18n).baseText("insights.dashboard.title")), 1)
          ]),
          _: 1
        }),
        createBaseVNode("div", null, [
          unref(insightsStore).isSummaryEnabled ? (openBlock(), createBlock(InsightsSummary, {
            key: 0,
            summary: unref(insightsStore).summary.state,
            loading: unref(insightsStore).summary.isLoading,
            class: normalizeClass(_ctx.$style.insightsBanner)
          }, null, 8, ["summary", "loading", "class"])) : createCommentVNode("", true),
          unref(insightsStore).isInsightsEnabled ? (openBlock(), createElementBlock("div", {
            key: 1,
            class: normalizeClass(_ctx.$style.insightsContent)
          }, [
            createBaseVNode("div", {
              class: normalizeClass(_ctx.$style.insightsChartWrapper)
            }, [
              unref(insightsStore).charts.isLoading ? (openBlock(), createElementBlock(Fragment, { key: 0 }, [
                createTextVNode(" loading ")
              ], 64)) : (openBlock(), createBlock(resolveDynamicComponent(chartComponents.value[props.insightType]), {
                key: 1,
                type: props.insightType,
                data: unref(insightsStore).charts.state
              }, null, 8, ["type", "data"]))
            ], 2),
            createBaseVNode("div", {
              class: normalizeClass(_ctx.$style.insightsTableWrapper)
            }, [
              createVNode(unref(InsightsTableWorkflows), {
                data: unref(insightsStore).table.state,
                loading: unref(insightsStore).table.isLoading,
                "onUpdate:options": fetchPaginatedTableData
              }, null, 8, ["data", "loading"])
            ], 2)
          ], 2)) : (openBlock(), createBlock(unref(InsightsPaywall), {
            key: 2,
            "data-test-id": "insights-dashboard-unlicensed"
          }))
        ])
      ], 2);
    };
  }
});
const insightsView = "_insightsView_1upqv_123";
const insightsBanner = "_insightsBanner_1upqv_133";
const insightsContent = "_insightsContent_1upqv_140";
const insightsChartWrapper = "_insightsChartWrapper_1upqv_149";
const insightsTableWrapper = "_insightsTableWrapper_1upqv_154";
const style0 = {
  insightsView,
  insightsBanner,
  insightsContent,
  insightsChartWrapper,
  insightsTableWrapper
};
const cssModules = {
  "$style": style0
};
const InsightsDashboard = /* @__PURE__ */ _export_sfc(_sfc_main, [["__cssModules", cssModules]]);
export {
  InsightsDashboard as default
};
